lib/rdc.js

```diff
@@ -0,0 +1,97 @@
+'use strict';
+
+const fs = require('fs');
+const path = require('path');
+
+const rdc = (source, destination, options, callback) => {
+  if (typeof callback !== 'function') {
+    if (typeof options !== 'function') throw new Error('ERR_INVALID_CALLBACK');
+    callback = options;
+    options = {};
+  }
+
+  const depth = options.depth;
+  const Transform = options.change.transform;
+  const readStreamOptions = options.change.readStreamOptions;
+  const writeStreamOptions = options.change.writeStreamOptions;
+  const filter = options.filter;
+  const stop = options.stopOnError;
+  const errorLog = options.errorLog;
+  const dereference = options.dereference;
+
+  const error = err => {
+    if (stop) throw new Error(err);
+    else if (errorLog) errorLog.write(JSON.stringify(err));
+    else callback(err);
+  };
+
+  const copyDirectory = target => {
+    fs.mkdirSync(target, err => {
+      if (err) error(err);
+    });
+  };
+
+  const copyFile = (current, target) => {
+    if (Transform) {
+      fs.createReadStream(current, readStreamOptions)
+        .pipe(new Transform())
+        .pipe(fs.createWriteStream(target, writeStreamOptions));
+    } else {
+      fs.copyFile(current, target, err => {
+        if (err) error(err);
+      });
+    }
+  };
+
+  const copyLink = (current, target) => {
+    fs.readlink(current, (err, linkString) => {
+      if (err) error(err);
+      if (dereference) {
+        copyDirectory(target);
+        start(path.resolve(path.dirname(current), linkString), target);
+      } else {
+        fs.symlink(linkString, target, err => {
+          if (err) error(err);
+        });
+      }
+    });
+  };
+
+  const start = (curDir = source, dest = destination, curDepth = 1) => {
+    if (curDepth > depth) return;
+    fs.readdir(curDir, (err, items) => {
+      if (err) error(err);
+      items.forEach(item => {
```

Comment:
It will be better to use [@metarhia/metasync](github.com/metarhia/metasync) for such kind of array processing.

Reply:
I don't understand what do you mean, should I use metasync for ```fs.readdir``` or ```for items.forEach``` or for both?